File: spin.ts
import chalk from "chalk"
import ora from "ora"

export async function spin<Type>(
    message: string,
    promise: Promise<Type>
): Promise<Type> {
    const spinner = ora(message).start()
    const result = await promise
    spinner.suffixText = chalk.bold("Done!")
    spinner.stopAndPersist()
    return result
}

File: lib/spin.dart
import 'dart:io';

import 'package:cli_spin/cli_spin.dart';
import 'package:io/ansi.dart';

Future<T> spin<T>(String message, Future<T> future) async {
  final spinner = CliSpin(text: message).start();
  try {
    final result = await future;
    spinner.success(green.wrap(message));
    return result;
  } catch (error, stack) {
    spinner.fail(red.wrap("An error occurred: $error\n\nStacktrace: $stack"));
    exit(0);
  }
}
